src/miner_lora.erl

```diff
@@ -181,48 +235,67 @@ handle(<<?PROTOCOL_2:8/integer-unsigned, Token:2/binary, ?TX_ACK:8/integer-unsig
                     {error, {unknown, Error}}
             end,
             gen_server:reply(From, Reply),
-            State#state{packet_timers = maps:remove(Token, State#state.packet_timers)};
+            State#state{packet_timers=maps:remove(Token, Timers)};
         error ->
             State
     end;
-handle(Packet, _IP, _Port, State) ->
-    lager:info("unhandled packet ~p", [Packet]),
+handle_udp_packet(Packet, _IP, _Port, State) ->
+    lager:info("unhandled udp packet ~p", [Packet]),
     State.
 
-handle_DATA([], _Gateway, State) ->
+-spec handle_json_data(list(), gateway(), state()) -> state().
+handle_json_data([], _Gateway, State) ->
     State;
-handle_DATA([{<<"rxpk">>, Packets}|Tail], Gateway, State) ->
-    State2 = handle_packet(sort_packets(Packets), Gateway, State),
-    handle_DATA(Tail, Gateway, State2);
-handle_DATA([{<<"stat">>, Status}|Tail], Gateway0, State) ->
-    Gateway = Gateway0#gateway{status=Status},
+handle_json_data([{<<"rxpk">>, Packets}|Tail], Gateway, State0) ->
+    State1 = handle_packets(sort_packets(Packets), Gateway, State0),
+    handle_json_data(Tail, Gateway, State1);
+handle_json_data([{<<"stat">>, Status}|Tail], Gateway0, #state{gateways=Gateways}=State) ->
+    Gateway1 = Gateway0#gateway{status=Status},
     lager:info("got status ~p", [Status]),
-    lager:info("Gateway ~p", [lager:pr(Gateway, ?MODULE)]),
-    handle_DATA(Tail, Gateway, State#state{gateways = maps:put(Gateway#gateway.mac, Gateway, State#state.gateways)}).
+    lager:info("Gateway ~p", [lager:pr(Gateway1, ?MODULE)]),
+    Mac = Gateway1#gateway.mac,
+    handle_json_data(Tail, Gateway1, State#state{gateways=maps:put(Mac, Gateway1, Gateways)}).
+
+-spec sort_packets(list()) -> list().
+sort_packets(Packets) ->
+    lists:sort(
+        fun(A, B) ->
+            proplists:get_value(<<"lsnr">>, A) >= proplists:get_value(<<"lsnr">>, B)
+        end,
+        Packets
+    ).
 
-handle_packet([], _Gateway, State) ->
+-spec handle_packets(list(), gateway(), state()) -> state().
+handle_packets([], _Gateway, State) ->
     State;
-handle_packet([Packet|Tail], Gateway, State) ->
+handle_packets([Packet|Tail], Gateway, #state{pubkey_bin=PubKeyBin, sig_fun=SigFun}=State) ->
     Data = base64:decode(proplists:get_value(<<"data">>, Packet)),
     case route(Data) of
         error ->
+            % TODO: Should we do the other packets?
```

Comment:
Doesn't it fall through after the case to the tail call to handle_packets?

Reply:
You are right I missed that